iree/compiler/Dialect/HAL/Transforms/CSEVariableLoads.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+class CSEVariableLoadsPass
+    : public PassWrapper<CSEVariableLoadsPass, OperationPass<FuncOp>> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<HALDialect>();
+  }
+
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+
+    Block *entryBlock = &funcOp.body().front();
+    auto entryBlockBuilder = OpBuilder::atBlockBegin(entryBlock);
+
+    DenseMap<StringRef, SmallVector<VariableLoadOp, 4>>
+        loadOpsGroupedByVariable;
+    funcOp.walk([&](VariableLoadOp loadOp) {
+      // Note: we assume that no IsolatedFromAbove regions are used with HAL
```

Comment:
You also will need to check for calls. If we were going to have this around longer we would want to do this like dispatchability analysis where we first go through each function and see what variables it writes and then aggregate that into a transitive set of all variables written by the function or any function that may be reached under it. With that information during the removal pass we can just check the analysis cache for the func and remove things that aren't in it (as we know that then the function and all reachable functions can't write it).

For now, we run the inliner and it's really aggressive so after inlining we aren't ending up with many calls besides the ones we prevent with noinline. At least bailing on a call will make this pass correct, and then it'll be easy to spot once we have some models with calls preserved.

Reply:
Checked for `CallOp` and bailed. Not analyzing more deeply [yet]. Thanks for the explanation!